media_driver/linux/common/os/i915/mos_bufmgr.c

```diff
@@ -957,27 +1042,58 @@ mos_gem_bo_alloc_internal(struct mos_bufmgr *bufmgr,
                 mos_gem_bo_free(&bo_gem->bo);
                 goto retry;
             }
+            if (mos_gem_bo_check_mem_region_internal(&bo_gem->bo, mem_type)) {
+                mos_gem_bo_free(&bo_gem->bo);
+                goto retry;
+            }
         }
     }
     pthread_mutex_unlock(&bufmgr_gem->lock);
 
     if (!alloc_from_cache) {
-        struct drm_i915_gem_create create;
 
         bo_gem = (struct mos_bo_gem *)calloc(1, sizeof(*bo_gem));
         if (!bo_gem)
             return nullptr;
 
         bo_gem->bo.size = bo_size;
-
-        memclear(create);
-        create.size = bo_size;
-
-        ret = drmIoctl(bufmgr_gem->fd,
+        bo_gem->mem_region = I915_MEMORY_CLASS_SYSTEM;
+
+        if(bufmgr_gem->has_lmem &&
+            (mem_type == MOS_MEMPOOL_VIDEOMEMORY || mem_type == MOS_MEMPOOL_DEVICEMEMORY)) {
+            struct drm_i915_gem_memory_class_instance mem_region;
+            memclear(mem_region);
+            mem_region.memory_class = I915_MEMORY_CLASS_DEVICE;
+            mem_region.memory_instance = 0;
+
+            struct drm_i915_gem_create_ext_memory_regions regions;
+            memclear(regions);
+            regions.base.name = I915_GEM_CREATE_EXT_MEMORY_REGIONS;
```

Comment:
uAPI has changed here vs. what we had in https://github.com/intel/media-driver/blob/master/media_driver/linux/common/os/i915_production/mos_bufmgr.c